lib/src/modules/apply.cc

```diff
@@ -0,0 +1,252 @@
+# Additionally include rundir, which gets its own initialization.
+puppet_root = Dir.mktmpdir
+moduledir = File.join(puppet_root, 'modules')
+Dir.mkdir(moduledir)
+
+# Use isolated directories in puppet_root *except* confdir so we can connect to puppetserver
+setting_keys = Puppet::Settings::REQUIRED_APP_SETTINGS - [:confdir] + [:rundir]
+cli = (setting_keys).flat_map do |setting|
+  ["--#{setting}", File.join(puppet_root, setting.to_s.chomp('dir'))]
+end
+cli << '--modulepath' << moduledir
+Puppet.initialize_settings(cli)
+
+# Avoid extraneous output
+Puppet[:report] = false
+
+# Make sure to apply the catalog, TODO (on orch side): pipe through _noop
+Puppet[:noop] = apply_opts['noop'] || false
+# TODO(on orch side): enable apply setting, currently this is only configurable from bolt.yaml in bolt
+if apply_opts['apply_settings']
+  apply_opts['apply_settings'].each do |setting, value|
+    Puppet[setting.to_sym] = value
+  end
+end
+
+exit_code = 0
+begin
+
+  # TODO: REMOVE THESE THEY ARE DEBUG LOGS
+  # Puppet::Util::Log.newdestination(:console)
+  # Puppet[:log_level] = :info
+
+  # Sync plugins to cache. Note that we are explicitly *not* using an isolated confdir
+  # because we need to be able to communicate with puppetserver 
+  plugin_cache = args['plugin_cache']
+  plugin_dest = File.join(plugin_cache, 'plugins')
+  pluginfact_dest = File.join(plugin_cache, 'pluginfacts')
+  Puppet[:plugindest] = plugin_dest
+  Puppet[:pluginfactdest] = pluginfact_dest
+  remote_env_for_plugins = Puppet::Node::Environment.remote(args['environment'])
+  downloader = Puppet::Configurer::Downloader.new(
+    "plugin",
+    Puppet[:plugindest],
+    Puppet[:pluginsource],
+    Puppet[:pluginsignore],
+    remote_env_for_plugins
+  )
+  downloader.evaluate
+
+  source_permissions = Puppet::Util::Platform.windows? ? :ignore : :use
+  plugin_fact_downloader = Puppet::Configurer::Downloader.new(
+    "pluginfacts",
+    Puppet[:pluginfactdest],
+    Puppet[:pluginfactsource],
+    Puppet[:pluginsignore],
+    remote_env_for_plugins,
+    source_permissions
+  )
+  plugin_fact_downloader.evaluate
+
+  # Append the newe paths to the load path (copying them over to the new vardir takes time and seems unneeded)
+  $LOAD_PATH << plugin_dest << pluginfact_dest
+
+  Puppet[:default_file_terminus] = :file_server
+  # This happens implicitly when running the Configurer, but we make it explicit here. It creates the
+  # directories we configured earlier.
+  Puppet.settings.use(:main)
+
+  # Given we set up an empty env, it will be 'production' in this case
+  # Note we *do not* want to set this to the environment we are applying the catalog for
+  env = Puppet.lookup(:environments).get('production')
+
+  # Ensure custom facts are available for provider suitability tests
+  facts = Puppet::Node::Facts.indirection.find(SecureRandom.uuid, environment: env)
+
+  report = if Puppet::Util::Package.versioncmp(Puppet.version, '5.0.0') > 0
```

Comment:
We know we're gonna have puppet > 6 in our environment

Reply:
lol yeah that was copy paste ;)